TR-29: Give appropriate error responses to auth failure

Jpbroadvision opened the pull request:
### Description

Giving appropriate responses to auth failure prior to resource release to the frontend. This may include auth failure due to expired or wrong token provided.

The intention is to override google provided error implementation messages with our own and to make sure the backend is communicating with the frontend seamlessly.


## Type of change

- [x] Authentication Refactoring

# How Has This Been Tested?

The **loginUser** controller was separated from **UserController**  controller as GET and used as the Rest endpoint for the entry of login to request for a resource by the frontend.

Below are the test results performed

1. Two Mockup tests are performed on the loginUser controller when a **valid token** and an **invalid token** is provided from the frontend


![backend test report](https://user-images.githubusercontent.com/60193444/231793521-a0d0d21c-cda0-4224-8dbd-263199e62f2b.PNG)

2. Mockup Test coverage for the UserController( using loginUser )

![test coverage](https://user-images.githubusercontent.com/60193444/231794131-e1dab686-9c0c-4cdd-a986-071fac50017d.PNG)

3. Below is a screenshot that is seen from the frontend after a successful google auth login by a user and a **valid token** is sent to the backend for validation and resource acquisition.

A True connection from backend to frontend if the **token is valid**

![valid auth](https://user-images.githubusercontent.com/60193444/231794578-0ea967a6-0562-405d-b5bc-7d00376de76a.PNG)

A True connection from backend to frontend if a **token is invalid**

![invalid auth](https://user-images.githubusercontent.com/60193444/231794807-2aa939e4-1a63-42d2-9323-ce844fda2303.PNG)


# Checklist:

- [x] Mockup Test performed on loginUser to validate the provided token.
- [x] Appropriate authentication response to frontend (401 and an error header) 
- [x] Update all PR issues

PaulKpantey committed:
chore: Separate the endpoints

PaulKpantey committed:
chore: assert that the token is correct

PaulKpantey committed:
docs: Add comment

PaulKpantey committed:
feat: How to handle errors at the commencement of auth process

PaulKpantey committed:
feat: extend to override functionality from AuthenticationException

PaulKpantey committed:
dependency: Add jsonwebtoken, jakarta.servlet and apache.tomcat dependencies

PaulKpantey committed:
feat: Add exceptionHandling().authenticationEntryPoint() to handle to override default errors

PaulKpantey committed:
feat: Add TTResponse for formatting messages

PaulKpantey committed:
feat: Add Actual rest endpoint for first entry

PaulKpantey committed:
feat: Add test for loginUser controller

PaulKpantey committed:
ignore: Ignore testReport/

PaulKpantey committed:
Merge branch 'dev' into TR-29

PaulKpantey committed:
chore: Merge dev branch and adjusted imports

knocksto commented on src/test/java/com/ttrides/turntablRides/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    void loginUser_whenProvidedUnauthorisedOrExpiredOrInvalidToken_ShouldReturnInvalid_tokenWith401Exception() throws Exception {
```
The test method name is too long. You also want to stick to a particular naming convention, preferably camel case when working with java. Simplify the method name to something a bit shorter like;
"testLoginUserWithInvalidTokenReturns401ExceptionAndInvalidTokenErrorMessage"

Jpbroadvision replied:
Noted. Resolved

knocksto commented on src/test/java/com/ttrides/turntablRides/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+
+        String content = (String) errorResult.getResponse().getHeaderValue("WWW-Authenticate");
+        assert content != null;
+        assertTrue(content.contains("invalid_token"));
```
assertEquals is more ideal in this case

Jpbroadvision replied:
Ok sure. But I used **assertTrue**  because the actual string returned includes other text which is needed. below is the error message body, out of which we assert whether it has an **invalid_token** 

 **Bearer error="invalid_token", error_description="An error occurred while attempting to decode the Jwt: Signed JWT rejected: Another algorithm expected, or no matching key(s) found", error_uri="https://tools.ietf.org/html/rfc6750#section-3.1"**
 
 Nevertheless, I have updated the test to use **assertEquals**

knocksto commented on src/test/java/com/ttrides/turntablRides/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    void loginUser_WhenProvidedCorrectToken_ShouldBeOkWith200ResponseAndExactResponseObject() throws Exception{
```
use camel case here instead and shorten the name to something like this; testLoginWithCorrectTokenReturnsExpectedResponse

knocksto commented on src/main/java/com/ttrides/turntablRides/controllers/UserController.java:
```diff
@@ -1,16 +1,21 @@
-    public String currentUser() {
-        return "Secured endpoint";
+    @GetMapping(value = "/login", produces = {MediaType.APPLICATION_JSON_VALUE})
+    @ResponseBody
```
@ResponseBody is not required when using @RestController

Jpbroadvision replied:
Noted

knocksto reviewed (commented):
Kindly resolve all comments on the PR

PaulKpantey committed:
refactor: Rename tests methods and change assertTrue to assertEquals

Renamed tests methods according to the PR corrections and changed assertTrue to assertEquals

PaulKpantey committed:
dependency: Add mongodb dependency

This was required in the models

PaulKpantey committed:
chore: Remove @ResponseBody from loginUser endpoint

PaulKpantey committed:
chore: Add new token for test and substring error body

Added a new token for test after the update. Moreover, to get the actual error message, there was a need to parse and substring the exact message needed.

PaulKpantey committed:
chore: Remove unused module

PaulKpantey committed:
chore: Remove unused module

knocksto commented on src/main/java/com/ttrides/turntablRides/controllers/UserController.java:
```diff
@@ -1,16 +1,19 @@
 
 @RestController
-@RequestMapping("/api/v1/login")
+@RequestMapping("/api/v1")
```
since this is a user controller lets add /user to the main endpoint. eg. /api/v1/user

knocksto commented on src/main/java/com/ttrides/turntablRides/controllers/UserController.java:
```diff
@@ -1,16 +1,19 @@
 
 
 @RestController
-@RequestMapping("/api/v1/login")
```
You took out the /login part of the endpoint but you still have your test using /login.

Jpbroadvision replied:
No please, it wasn't taken out. I just separated the RequestMapping from GetMapping so, we can join other endpoints to the main user request later.

Because we would have smaller endpoints for the user later, I broke that **/login** just for the test [ @GetMapping(value = "/login") ]

knocksto commented on src/test/java/com/ttrides/turntablRides/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Test
+    void testLoginUserWithInvalidTokenReturns401ExceptionAndInvalidTokenErrorMessage() throws Exception {
+        MvcResult errorResult = mockMvc.perform(get("/api/v1/login")
```
your user controller doesn't have a /login route. Let's be consistent in the route. Please change it for the other occurrences in this class. Refer to the base endpoint of the UserController

Jpbroadvision replied:
Yeah, It does. @RequestMapping("/api/v1") maps it to @GetMapping(value = "/login", produces = {MediaType.APPLICATION_JSON_VALUE}) to form **/api/v1/login** earlier.

Moreover, the test needs the full endpoint, so it needed to be added (RequestMapping joined to GetMapping)

But I have updated it accordingly

PaulKpantey committed:
chore: Update endpoint and token for new test

PaulKpantey committed:
chore: Add /user to @RequestMapping
